Speed-up the concretizer with static analysis

alalazo opened the pull request:
Currently, when we setup the ASP problem for `clingo`, we don't take into account the configuration. This results in setting up ASP problems that are larger than necessary, with possibly redundant information, and higher concretization times. 

This PR tries to improve things by adding an opt-in feature that computes the _possible dependencies_ of a solve taking also into account the current configuration, and avoids adding possible dependencies that we are certain can't be in the final solution.

The feature can be activated with:
```yaml
concretizer:
  static_analysis: true
```

Examples of simple rules to discard dependencies are:
- Dependencies that are not buildable, and for which no binary is present (e.g. `cray-mpich` etc. on non Cray systems)
- Dependencies that are not for the current platform (e.g. `msmpi` on non Windows platforms)
- Conditional dependencies that cannot be activated, because of some user requirement (e.g. `cuda` etc. if the user requires `~cuda` in configuration)
- Virtual providers that cannot be used, because of a requirement on a virtual

The speed-up these rules seem to give depends on the use case at hand, but if the configuration is updated properly, they are noticeable. 

Since in cases where there is no rule to exclude packages upfront, reuse is active, and this option is activated, it's possible to see some minor slow down, the feature has been added as opt-in, so it's turned off by default.

### Modifications

- [x] Renamed `spack.solver.counter` to `spack.solver.input_analysis`
- [x] Add static analysis logic to create a reduced ASP problem
- [x] Remove `spack_package_base.possible_dependencies`
- [x] Remove `PackageBase.possible_dependencies`

<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

alalazo committed:
possible_dependencies: clean type-hints, make args keyword only

alalazo committed:
Create a new object to delegate searching for possible deps

alalazo committed:
Inlined the call to package_base.possible_dependencies

alalazo committed:
Inlined the call to PackageBase.possible_dependencies

alalazo committed:
Add a Context class, move runtime pkgs computation there

alalazo committed:
Deal with runtime packages in the analyzer

alalazo committed:
Delegate providers_for and is_virtual operations to contexts

alalazo committed:
Filter packages that are not for this platform

alalazo committed:
Proceed on virtuals that have been visited already

alalazo committed:
Filter on whether a package can be installed

alalazo committed:
Add an option to consider only preferred providers

alalazo committed:
Improve default packages.yaml configuration

alalazo committed:
Add more flags to possible_dependencies

alalazo committed:
Don't emit facts for deps we statically know cannot be added

alalazo commented:
This is an example test run on:

* **Spack:** 0.24.0.dev0 (b2a75db030cf4ba6580d980f6d10ea5aa85029fb)
* **Python:** 3.13.0
* **Platform:** linux-ubuntu20.04-icelake

using the following configuration:
```yaml
packages:
  all:
    require:    
    - "~cuda"
    - "~rocm"
    - "~qtgui"
    - "~guile"
```
to simulate the need to build software for the CPU, and to tell Spack that I don't want `qt` support for `cmake` or `guile` support for `gmake`. The raw data is:

[radiuss.txt](https://github.com/user-attachments/files/18557303/radiuss.txt)
[radiuss-develop-required.csv](https://github.com/user-attachments/files/18557306/radiuss-develop-required.csv)
[radiuss-branch-required.csv](https://github.com/user-attachments/files/18557307/radiuss-branch-required.csv)
[radiuss-branch-preferred-required.csv](https://github.com/user-attachments/files/18557308/radiuss-branch-preferred-required.csv)

![radiuss-required](https://github.com/user-attachments/assets/c227b81b-bad4-474a-a1f0-23cdd07d7516)

The `branch+preferred` column is obtained turning on the option to only allow providers that are explicitly mentioned in the preferred section in config.

alalazo committed:
pipelines: update e4s config to specify more variants

alalazo committed:
Move Context to its own file

alalazo committed:
Add Context.candidate_targets

alalazo committed:
Move content of static.py into counter.py

alalazo committed:
Optimize on targets, if no package can use one

alalazo committed:
e4s: further specify spack.yaml

alalazo committed:
developer-tools: more specific spack.yaml

alalazo commented:
As a point example:
- https://gitlab.spack.io/spack/spack/-/jobs/14843321 (e4s concretized on this branch, on uo-mothra)
- https://gitlab.spack.io/spack/spack/-/jobs/14826616 (e4s concretized on develop, on uo-mothra)

This branch is roughly 20% faster (thanks @eugeneswalker for the data and links!)

alalazo commented:
> what if we also limited rendering CPU micro-architectures when a specific architecture is required?

Done. It has a much smaller effect than providers, will try to measure it and post.

alalazo committed:
Revert changes in package_base.py

alalazo committed:
Fix docstrings and type-hints

alalazo committed:
Add buildcache to exclusion logic

haampie commented on lib/spack/spack/solver/counter.py:
```diff
@@ -177,3 +164,144 @@ def possible_packages_facts(self, gen, fn):
         gen.newline()
+
+
+class _PossibleDependenciesAnalyzer:
```
Maybe move it out of here and drop `PackageBase.possible_dependencies` then? It has just 2 call sites. Also thinking to replace recursion by a stack, since it's pretty trivial and that makes the stacktraces a bit easier to follow in `viztracer`/perfetto. If you do that, you can get rid of _possible_dependencies, after which you can get rid of `class _PossibleDependenciesAnalayzer`, and it's just a function.

~Looks like avoiding recursion is ±2.2x faster, but only if package classes were already loaded, which remaines the bottleneck of course.~ edit: nope, with recursion GC just happens to trigger more often, if not it's roughly equal.

alalazo commented:
Results for the same platform as before, using:
```yaml
packages:
  all:
    require:    
    - "~cuda"
    - "~rocm"
    - "~qtgui"
    - "~guile"
    - target=x86_64_v3
```

[radiuss-develop-required-target.csv](https://github.com/user-attachments/files/18621948/radiuss-develop-required-target.csv)
[radiuss-branch-required-preferred-target.csv](https://github.com/user-attachments/files/18621951/radiuss-branch-required-preferred-target.csv)
[radiuss-branch-required-target.csv](https://github.com/user-attachments/files/18621952/radiuss-branch-required-target.csv)

![radiuss-required-target](https://github.com/user-attachments/assets/7017dcab-2918-48b8-84ce-d6eb29c83c95)

alalazo committed:
Use a stack for _ possible_dependencies

alalazo committed:
Further cleanup

haampie commented on lib/spack/spack/solver/counter.py:
```diff
@@ -199,109 +199,106 @@ def possible_dependencies(
+        while stack:
+            pkg_name = stack.pop()
+
+            visited.setdefault(pkg_name, set())
```
Bit unclear why `packages` are followed but children are not.

Maybe just `if pkg_name in visited: continue; else visited[pkg_name] = set()`? then you dedupe `packages` too. I would keep `if dep_name in visited:` below too because the number of leaf nodes is of the order of the graph size.

alalazo replied:
> Bit unclear why packages are followed but children are not.

I didn't change the implementation here, but I think this structure is there to deal with the `transitive=False` case. I'll try to simplify, or split, it.

alalazo replied:
Done

haampie commented on lib/spack/spack/solver/counter.py:
```diff
@@ -199,109 +199,106 @@ def possible_dependencies(
 
-                if not self.context.is_allowed_on_this_platform(pkg_name=dep_name):
-                    continue
+                    visited.setdefault(dep_name, set())
```
```visited[dep_name] = set()```

alalazo replied:
If we go with your suggestion above, I think this one needs to be deleted, not changed this way. Otherwise
```
if pkg_name in visited:
```
will be true without the package being visited. I didn't change the previous implementation, but I agree that it can be simplified.

haampie commented on lib/spack/spack/solver/counter.py:
```diff
@@ -199,109 +199,106 @@ def possible_dependencies(
-            for dep_name in dep_names:
-                if dep_name in visited:
-                    continue
+                visited.setdefault(pkg_cls.name, set()).update(dep_names)
```
why is pkg_cls.name used here? is it different from pkg_name?

alalazo replied:
No, this one was not caught by the refactoring tool. I'll fix it

alalazo replied:
Fixed in ee2c25500e9a72a81289c8aa04c2ff2883b05394

alalazo committed:
pkg_cls.name -> pkg_name

alalazo committed:
Simplify deptype conditions

alalazo committed:
Extract a method to turn input into a list of str

alalazo committed:
Exit early in the while loop

To account for transitive = False, we need to
keep track of depth, if we don't want to change
the implementation considerably

alalazo committed:
Remove depth, use edges to reconstruct direct deps

alalazo committed:
Revert changing defaults in packages.yaml

alalazo committed:
Remove spack.package_base.possible_dependencies

alalazo committed:
Remove PackageBase.possible_dependencies

alalazo committed:
Cleanup tests

haampie reviewed (changes requested):
In practice I only see a slowdown for `spack solve --timers gcc` with a reasonably sized database

```
PR setup          5.457s
develop setup     4.937s
```

solve time is the same.

I would also not add yet another config option `preferred_providers_only` since we would have to continue to support it and i foresee it will bit rot. The equivalent is package requirements.

alalazo committed:
Inject globals instead of reconstructing them

This introduces the risk of injecting objects
that are not synced, but avoids fs operations
done on a newly created store / repo.

alalazo committed:
Split a dataclass out of the class performing analysis

alalazo committed:
Make the binary index part of Context

alalazo committed:
Make static analysis opt-in

alalazo committed:
Remove unneeded class

alalazo committed:
pipelines: update

alalazo committed:
pipelines: update

alalazo committed:
Compute libc pkgs in constructor

alalazo committed:
Remove duplicate

alalazo committed:
Make the config option a simple bool, for now

Can be extended later, if need be

alalazo commented:
Note that this option might be useful also in simpler scenarios, like:
```console
$ spack solve --timers gcc
```
without having the 0.5s slowdown you show above, if we ship sensible defaults per platform, in particular after we merge #45189 (excluding by default uncommon compilers might help speeding up concretization). 

Anyhow, I'll follow with other PRs to change the default, and improve concretization speed - for now I think it's good if we merge this as an opt-in feature.

alalazo committed:
_default_unreachable: deduplicate code

alalazo committed:
PossibleDependencyGraph: extract an interface

alalazo committed:
Rename PossibleDependencyGraph implementations

alalazo committed:
Remove the Context class

alalazo committed:
Put everything in a single module

alalazo committed:
Move factory function

alalazo commented:
@haampie The refactor that we talked about in DM should be done:
- Removed the `Context` class
- Have an interface for `PossibleGraph`
- Renaming the classes doing static analysis (or not)

I also put both counters and the new classes in a module called `spack.solver.input_analysis`.

haampie commented on lib/spack/spack/solver/input_analysis.py:
```diff
@@ -0,0 +1,524 @@
+    edges: Dict[str, Set[str]]
+
+
+class PossibleDependencyGraph:
```
What I said was to have a base class with an implementation for traversal of the graph, and either subclass it or do a visitor pattern for static analysis where you don't follow certain edges. Is that not feasible? Now `possible_dependencies` is coupled to "static analysis", and calling `create_graph_analyzer` from `graph.py` and `cmd/dependencies.py` is incorrect since they should be config independent? Or is that intentional?

`unreachable`/`candidate_targets` don't belong in the class that does traversal, it belongs to the analysis stuff, and I think your interface is an interface for analysis, not for traversal.

alalazo replied:
> Now possible_dependencies is coupled to "static analysis", and calling create_graph_analyzer from graph.py and cmd/dependencies.py is incorrect since they should be config independent? Or is that intentional?

It's intentional, since it's a quick way to show what is considered during the solve, given the current configuration. For instance, with:
```yaml
packages:
  mpi:
    require: mpich
  java:
    require: openjdk
  zlib-api:
    require: zlib-ng+compat
```
you should get something like:
```
$ spack -c concretizer:static_analysis:false dependencies -t hdf5 | wc -l
563
$ spack -c concretizer:static_analysis:true dependencies -t hdf5 | wc -l
513

```

alalazo replied:
> I think your interface is an interface for analysis, not for traversal.

Correct

haampie reviewed (approved):
Some design improvements can be done as a follow-up.
